common/appscale/common/service_stats/stats_manager.py

```diff
@@ -0,0 +1,492 @@
+    "4xx": metrics.CountOf(matchers.CLIENT_ERROR),
+    "5xx": metrics.CountOf(matchers.SERVER_ERROR),
+    "avg_latency": metrics.Avg("latency")
+  }
+}
+# This metrics map corresponds to following output:
+# {
+#   "from": 1515699718987,
+#   "to": 1515735126789,
+#   "all": 1225,
+#   "4xx": 11,
+#   "5xx": 3,
+#   "avg_latency": 325,
+#   "by_app": {
+#     "guestbook": {
+#       "by_resource": {
+#         "/get/user": {"all": 56, "4xx": 11, "5xx": 3, "avg_latency": 321},
+#         "/": {"all": 300, "4xx": 0, "5xx": 0, "avg_latency": 68}
+#       },
+#       "all": 356,
+#       "4xx": 11,
+#       "5xx": 3,
+#       "avg_latency": 107
+#     },
+#     "validity": {
+#       "by_resource": {
+#         "/wait": {"all": 69, "4xx": 0, "5xx": 0, "avg_latency": 5021},
+#         "/version": {"all": 800, "4xx": 0, "5xx": 0, "avg_latency": 35}
+#       },
+#       "all": 869,
+#       "4xx": 0,
+#       "5xx": 0,
+#       "avg_latency": 430
+#     }
+#   }
+# }
+
+
+class ServiceStats(object):
+  """
+  Collects stats about requests handled by a service.
+  Service should report any request it starts working on using:
+    start_request(request_info_dict)
+  After finishing request service should call:
+    finalize_request(request_no, result_info_dict)
+  (its recommended to do this in finally section of try-except statement).
+
+  To retrieve cumulative stats call:
+    get_cumulative_stats()
+  To retrieve statistics for recent requests use one of:
+    get_recent(age)
+    scroll_recent(cursor)
+  """
+
+  DEFAULT_HISTORY_SIZE = 1000
+
+  DEFAULT_MAX_REQUEST_AGE = 60 * 60 * 2  # Force clean requests older than 2h
+  AUTOCLEAN_INTERVAL = 60 * 60 * 4
+
+  RESERVED_REQUEST_FIELDS = ["start_time", "end_time", "latency"]
+
+  def __init__(self, service_name, history_size=DEFAULT_HISTORY_SIZE,
+               force_clean_after=DEFAULT_MAX_REQUEST_AGE,
+               request_fields=DEFAULT_REQUEST_FIELDS,
+               cumulative_counters=DEFAULT_CUMULATIVE_COUNTERS,
+               default_metrics_for_recent=SINGLE_APP_METRICS_MAP):
+    """ Initialises an instance of ServiceStats.
+
+    Args:
+      service_name: a str representing name of service.
+      history_size: a number of recent requests to store.
+      force_clean_after: seconds to wait before removing started requests.
+      request_fields: a list of request property names.
+      cumulative_counters: a dictionary describing cumulative counters config.
+      default_metrics_for_recent: a dictionary containing metrics
+        which should be computed for recent requests by default.
+    """
+
+    now = _now()
+    self._service_name = service_name
+    self._request_info_class = self.generate_request_model(request_fields)
+
+    # Initialize properties for tracking latest N requests
+    self._last_request_no = 0
+    self._current_requests = {}  # {request_no: RequestInfo()}
+    self._finished_requests = []  # circular list containing recent N requests
```

Comment:
@cdonati , thanks, I didn't know about this module, though guessed that there should be something like this. I consciously didn't search for a such tool, because manual limiting length introduces only 2 new lines without big performance impact.

Now I ran a performance test using timeit and 1 000 000 items added to the list and deque with a size of 1000 elements. Result is 0.25 vs 0.41 seconds, deque wins, but adding new import and loosing ability to use bisect is worse in my mind.

Reply:
You're right. I don't think it matters at 1000 elements. It would only become a problem if we decide to keep more.